traits/src/block_median_time_context.rs

```diff
@@ -1,24 +1,47 @@
 use ckb_core::BlockNumber;
+use numext_fixed_hash::H256;
+use std::cmp::min;
 
 /// The invoker should only rely on `block_median_time` function
 /// the other functions only use to help the default `block_median_time`, and maybe unimplemented.
 pub trait BlockMedianTimeContext {
     fn median_block_count(&self) -> u64;
-    /// block timestamp
-    fn timestamp(&self, block_number: BlockNumber) -> Option<u64>;
-    /// ancestor timestamps from a block
-    fn ancestor_timestamps(&self, block_number: BlockNumber) -> Vec<u64> {
-        let count = self.median_block_count();
-        (block_number.saturating_sub(count.saturating_sub(1))..=block_number)
-            .filter_map(|n| self.timestamp(n))
-            .collect()
-    }
 
-    /// get block median time
-    fn block_median_time(&self, block_number: BlockNumber) -> Option<u64> {
-        let mut timestamps: Vec<u64> = self.ancestor_timestamps(block_number);
-        timestamps.sort_by(|a, b| a.cmp(b));
+    /// Return timestamp of the correspoding bloch_hash, and hash of parent block
+    ///
+    /// Fake implementation:
+    /// ```ignore
+    /// let current_header = get_block_header(block_hash);
+    /// let parent_header = current_header.timestamp_and_parent().header();
+    /// return (parent_header.timestamp(), parent_header.hash());
+    /// ```
+    fn timestamp_and_parent(&self, block_hash: &H256) -> (u64, H256);
+
+    /// Return past block median time, **including the timestamp of the given one**
+    ///
+    /// It's duty for outside caller to ensure that the block_number and block_hash are matched
+    fn block_median_time(&self, block_number: BlockNumber, block_hash: &H256) -> u64 {
+        let median_time_span = min(block_number + 1, self.median_block_count());
```

Comment:
Do we really need the block_number argument?

Reply:
Without `block_number`, we must assume the parent of genesis block is "0x0000000000000".
If we accept this assumption, it's ok to remove this parameter here, in another PR.
I am not sure about whether we should make this assumption.

Reply:
IMO it is more like a part of consensus rule rather than an assumption.
However, we are missing some part to verify special rules for the genesis block.